File: apps/server-py/src/agents/agent_registry.py
from src.agents.agent_types import AgentType
from src.lib.logger import logger

_registry: dict[AgentType, object] = {}


def register_agent(agent: object) -> None:
    if agent.type in _registry:  # type: ignore[attr-defined]
        logger.warning(f"Agent already registered — overwriting: {agent.type}")  # type: ignore[attr-defined]
    _registry[agent.type] = agent  # type: ignore[attr-defined]
    logger.info(f"Agent registered: {agent.type} ({agent.display_name})")  # type: ignore[attr-defined]


def get_agent(agent_type: AgentType) -> object:
    agent = _registry.get(agent_type)
    if not agent:
        raise ValueError(
            f'No agent registered for type "{agent_type}". '
            f"Registered: {list(_registry.keys())}"
        )
    return agent


def list_agents() -> list[object]:
    return list(_registry.values())

File: apps/server/src/agents/agent.registry.ts
import { AgentType, IAgent } from "./agent.types";
import { logger } from "../lib/logger";

// simple in-memory registry — agents are registered once at startup
const registry = new Map<AgentType, IAgent>();

export function registerAgent(agent: IAgent): void {
  if (registry.has(agent.type)) {
    logger.warn("Agent already registered — overwriting", { type: agent.type });
  }
  registry.set(agent.type, agent);
  logger.info("Agent registered", { type: agent.type, name: agent.displayName });
}

export function getAgent(type: AgentType): IAgent {
  const agent = registry.get(type);
  if (!agent) {
    throw new Error(
      `No agent registered for type "${type}". Registered: [${[...registry.keys()].join(", ")}]`,
    );
  }
  return agent;
}

export function listAgents(): IAgent[] {
  return [...registry.values()];
}
